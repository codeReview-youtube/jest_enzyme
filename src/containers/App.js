import React, { Component } from 'react';
import { Items, InputField } from '../components';
import fetch from 'node-fetch';

export default class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      todos: [],
      todo: '',
      id: null
    };
  }

  componentDidMount() {
    this.fetchTodos();
  }
  /**
   * onChange
   */
  onChange = newTodo => {
    this.setState(() => ({
      todo: newTodo
    }));
  };
  /**
   * onSubmit
   */
  onSubmit = () => {
    const { todo, todos, id } = this.state;
    if (todo.trim() !== '') {
      // if id exist => it's update
      if (id) {
        todos[id - 1].title = todo;
        todos[id - 1].completed = false;
        this.setState(() => ({
          todos,
          id: null
        }));
      } else {
        // otherwise add
        this.setState(() => ({
          todos: [
            {
              id: todos.length + 1,
              completed: false,
              title: todo
            },
            ...todos
          ]
        }));
      }
    } else {
      return null;
    }

    this.clear();
  };
  /**
   * onEdit
   */
  onEdit = INDEX => {
    const { todos } = this.state;
    this.setState(() => ({
      id: todos.find((_, i) => i === INDEX).id,
      todo: todos.find((_, i) => i === INDEX).title
    }));
  };
  /**
   * clear
   */
  clear = () => {
    this.setState(() => ({
      todo: ''
    }));
  };
  /**
   * onCheck
   */
  onClick = index => {
    const { todos } = this.state;
    todos[index].completed = !todos[index].completed;
    this.setState({
      todos
    });
  };
  /**
   * onDelete
   */
  onDelete = index => {
    const { todos } = this.state;
    this.setState(() => ({
      todos: todos.filter((_, i) => i !== index)
    }));
  };
  /**
   * fetchTodos
   */
  fetchTodos() {
    return fetch('https://jsonplaceholder.typicode.com/todos')
      .then(rs => rs.json())
      .then(
        todos =>
          todos &&
          todos.length > 0 &&
          this.setState({
            todos
          })
      );
  }

  render() {
    const { todos, todo } = this.state;
    return (
      <div className='app'>
        <InputField
          onChange={this.onChange}
          value={todo}
          onSubmit={this.onSubmit}
        />
        <Items
          items={todos}
          onClick={this.onClick}
          onDelete={this.onDelete}
          onEdit={this.onEdit}
        />
      </div>
    );
  }
}
