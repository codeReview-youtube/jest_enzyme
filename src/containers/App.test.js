/**
 * @jest-environment jsdom
 */
/* eslint-disable react/display-name */
import React from 'react';
import { shallow } from 'enzyme';
/**
 *import  Component
 */
import App from './App';
/**
 * props
 */
const props = {};
/**
 * mock Todos
 */
const todos = [
  {
    id: 1,
    title: '',
    completed: false
  },
  {
    id: 2,
    title: '',
    completed: true
  }
];
/**
 * Tests: App Component
 */
describe('App Component', () => {
  /**
   * Snapshot:[default]
   */
  test('should match default snapshot', () => {
    expect.assertions(1);

    const wrapper = shallow(<App {...props} />);

    expect(wrapper).toMatchSnapshot();
  });
  /**
   * Snapshot:[with state]
   */
  test('should match with items', async () => {
    expect.assertions(1);

    const wrapper = shallow(<App {...props} />);

    await wrapper.instance().fetchTodos();
    // jsonplace holder return 200 todos
    expect(wrapper.state('todos')).toHaveLength(200);
  });
  /**
   * triggers [onDelete]
   */
  test('should delete completed todo', () => {
    expect.assertions(1);
    const wrapper = shallow(<App />);

    wrapper.setState({ todos });
    wrapper.instance().onDelete(1);

    expect(wrapper.state('todos')).toEqual([todos[0]]);
  });
  /**
   * triggers [onClick]
   */
  test('should change completed status', () => {
    expect.assertions(1);
    const wrapper = shallow(<App />);
    const index = 0;
    wrapper.setState({ todos });
    wrapper.instance().onClick(index);

    expect(wrapper.state('todos')[index].completed).toEqual(
      todos[index].completed
    );
  });
  /**
   * triggers [onEdit]
   */
  test('should change title with newTitle', () => {
    expect.assertions(2);

    const wrapper = shallow(<App />);
    const index = 0;

    wrapper.setState({ todos, todo: 'something', id: index + 1 });
    wrapper.instance().onClick(index);
    wrapper.instance().onSubmit();

    const { completed, title } = wrapper.state('todos')[index];
    expect(completed).toEqual(todos[index].completed);

    expect(title).toEqual('something');
  });
  /**
   * triggers [onChange]
   */
  test('should add new Todo', () => {
    expect.assertions(1);
    const wrapper = shallow(<App />);
    const newTodo = 'something';

    wrapper.instance().onChange(newTodo);
    wrapper.instance().onSubmit();

    expect(wrapper.state('todos')).toHaveLength(1);
  });
  /**
   * triggers [onEdit]
   */
  test('should trigger onEdit to setState', () => {
    expect.assertions(1);
    const wrapper = shallow(<App />);
    const index = 0;
    wrapper.setState({ todos });
    wrapper.instance().onEdit(index);

    expect(wrapper.state('id')).toEqual(index + 1);
  });
  /**
   * triggers [onSubmit]
   */
  test('should return null if value [""]', () => {
    expect.assertions(1);
    const wrapper = shallow(<App />);

    const value = wrapper.instance().onSubmit();

    expect(value).toEqual(null);
  });
});
