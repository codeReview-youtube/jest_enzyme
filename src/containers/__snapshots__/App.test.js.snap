// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Component should match default snapshot 1`] = `
<div
  className="app"
>
  <InputField
    onChange={[Function]}
    onSubmit={[Function]}
    value=""
  />
  <Items
    items={Array []}
    onClick={[Function]}
    onDelete={[Function]}
    onEdit={[Function]}
  />
</div>
`;
